snuba/optimize.py

```diff
@@ -31,31 +34,153 @@ def _get_metrics_tags(table: str, clickhouse_host: Optional[str]) -> Mapping[str
+            cutoff_time=full_job_end_time,
+        )
+
+        if start_time < parallel_start_time:
+            return [pre_parallel_bucket, parallel_bucket, final_bucket]
+        elif start_time < parallel_end_time:
+            return [parallel_bucket, final_bucket]
+        else:
+            return [final_bucket]
+
+
 def run_optimize(
     clickhouse: ClickhousePool,
     storage: ReadableTableStorage,
     database: str,
     before: Optional[datetime] = None,
-    ignore_cutoff: bool = False,
-    parallel: int = 1,
-    clickhouse_host: Optional[str] = None,
 ) -> int:
+    """
+    The most basic form of running an optimize final on a storage.
+
+    It is deliberately simple and separate from `run_optimize_cron_job` since
+    there is a need to be able to run optimizations from the replacer.
+    """
+    schema = storage.get_schema()
+    assert isinstance(schema, TableSchema)
+    table = schema.get_local_table_name()
+    database = storage.get_cluster().get_database()
+
+    partitions = get_partitions_to_optimize(
+        clickhouse, storage, database, table, before
+    )
+    partition_names = [partition.name for partition in partitions]
+
+    optimize_partitions(
+        clickhouse=clickhouse,
+        database=database,
+        table=table,
+        partitions=partition_names,
+    )
+
+    return len(partitions)
+
+
+def run_optimize_cron_job(
+    clickhouse: ClickhousePool,
+    storage: ReadableTableStorage,
+    database: str,
+    parallel: int,
+    clickhouse_host: str,
+    tracker: OptimizedPartitionTracker,
+    before: Optional[datetime] = None,
+) -> int:
+    """
+    The sophisticated form of running an optimize final on a storage.
+
+    The sophistication include:
+    1. Being able to run multiple optimize final jobs concurrently.
+    2. Being able to set time boundaries for different phases of optimization.
+    3. Being able to work with stateful optimization states.
+    """
     start = time.time()
     schema = storage.get_schema()
     assert isinstance(schema, TableSchema)
     table = schema.get_local_table_name()
     database = storage.get_cluster().get_database()
 
-    parts = get_partitions_to_optimize(clickhouse, storage, database, table, before)
+    optimization_bucket = _build_optimization_buckets(
+        start_time=datetime.now(), max_parallel=parallel
+    )
+
+    try:
+        partitions_to_optimize = tracker.get_partitions_to_optimize()
+    except NoOptimizedStateException:
+        # We don't have any recorded state of partitions needing optimization
+        # for today. So we need to build it.
+        partitions = get_partitions_to_optimize(
+            clickhouse, storage, database, table, before
+        )
+        if len(partitions) == 0:
+            logger.info("No partitions need optimization")
+            return 0
+        partition_names = [partition.name for partition in partitions]
+        tracker.update_all_partitions(partition_names)
```

Comment:
What happens if the job gets restarted past the daily job cutoff time? Like 23:11.
The TTL we set on Redis will be in the past. Will it be ok ?